heareval/tasks/speech_commands.py

```diff
@@ -82,50 +184,182 @@ def run(self):
         )
 
         # Get label for the data from anywhere.
         # In this case it is the folder name
         process_metadata["label"] = process_metadata["relpath"].apply(
-            lambda relative_path: os.path.basename(os.path.dirname(relative_path))
+            ConfigureTrainValMetaData.apply_label
         )
 
         # Save the process metadata
         process_metadata.to_csv(
-            os.path.join(self.workdir, "process_metadata.csv"),
+            os.path.join(self.workdir, self.outfile),
             columns=PROCESSMETADATACOLS,
             header=False,
             index=False,
         )
 
-        with self.output().open("w") as _:
-            pass
+        self.mark_complete()
 
 
-class SubsampleCorpus(SubsampleCorpus):
+class CombineMetaData(WorkTask):
+    """
+    Because we dealt with the metadata for the train/val and testing datasets
+    separately this combines them together into a single csv file.
+    """
+
+    outfile = luigi.Parameter()
+
+    def requires(self):
+        return {
+            "train": ConfigureTrainValMetaData(outfile="train_corpus_metadata.csv"),
+            "test": ConfigureTestMetaData(outfile="test_corpus_metadata.csv"),
+        }
+
+    def run(self):
+        train = self.requires()["train"]
+        train_meta = pd.read_csv(
+            os.path.join(train.workdir, train.outfile),
+            header=None,
+            names=PROCESSMETADATACOLS,
+        )
+
+        test = self.requires()["test"]
+        test_meta = pd.read_csv(
+            os.path.join(test.workdir, test.outfile),
+            header=None,
+            names=PROCESSMETADATACOLS,
+        )
+
+        process_metadata = train_meta.append(test_meta)
+        # Save the process metadata
+        process_metadata.to_csv(
+            os.path.join(self.workdir, self.outfile),
+            columns=PROCESSMETADATACOLS,
+            header=False,
+            index=False,
+        )
+
+        self.mark_complete()
+
+
+class SubsamplePartition(SubsamplePartition):
+    """
+    A subsampler that acts on a specific partition.
+    All instances of this will depend on the combined process metadata csv.
+    """
+
     def requires(self):
         # The meta files contain the path of the files in the data
         # so we dont need to pass the extract as a dependency here.
-        return {"meta": ConfigureProcessMetaData()}
+        return {
+            "meta": CombineMetaData(outfile="process_metadata.csv"),
+        }
 
 
-class MonoWavTrimCorpus(MonoWavTrimCorpus):
+class SubsamplePartitions(WorkTask):
+    """
+    Aggregates subsampling of all the partitions into a single task as dependencies.
+    All the subsampled files are stored in the requires workdir, so we just link to
+    that since there aren't any real outputs associated with this task.
+    This is a bit of a hack -- but it allows us to avoid rewriting
+    the Subsample task as well as take advantage of Luigi concurrency.
```

Comment:
This felt a bit hacky with the symlink on the entire workdir -- but maybe it is okay

Reply:
Yes. I think it is okay. We are doing a similar thing for the resampling as well, so this looks consistent solution in our script for running things in parallel.